### ISSUE TITLE
Try.ToEither is called with null for exception

### ISSUE THREAD

**Issue by PepijnK** [louthy/language-ext#615]
Hi,
When a Try chain is filtered to a BottomException (no results) the ToEither call is called with a `null` for Exception value.
Not sure if this bad behavior or misuse of the ToEither call. Here is the part of the chain that triggers an Exception if executed (asynchronously):

```
                Try(() => new ItemSearch
                {
                   itemId=input
                })
                .ToAsync()
                .Bind(t => _ops.Search(t))
                .Select(searchResponse => searchResponse.searchResult)
                .Filter(searchResult => searchResult.status.isSuccess && searchResult.recordList != null &&
                                        searchResult.recordList.Length >= 1)
                .Select(searchResult =>
                {
                    return new Converted(searchResult)
                })
                .ToEither(ex => _logger.LogDebug("ex is null, while it should be BottomException!?"))
                .ToAsync();
```

**Comment by bender2k14** [comment]
Can you provide an example that compiles?

**Comment by gwintering** [comment]
These tests both pass
```csharp
        [Fact]
        public void BottomToEitherGivesException()
        {
            var tryBottom = Try(1).Filter(_ => false);
            Assert.True(tryBottom.Try().IsBottom, "Try not in Bottom (test requirement)");
            tryBottom.ToEither(Assert.IsType<BottomException>);
        }

        [Fact]
        public async Task AsyncBottomToEitherGivesException()
        {
            var tryBottom = TryAsync(1).Filter(_ => false);
            Assert.True((await tryBottom.Try()).IsBottom, "Try not in Bottom (test requirement)");
            await tryBottom.ToEither(Assert.IsType<BottomException>);
        }
```

**Comment by PepijnK** [comment]
Add another `select` after filter to the monadic chain and it will fail:

  ```
        [Fact]
        public void BottomToEitherGivesException()
        {
            var tryBottom = Try(1)
                .Filter(result => false)
                .Select(_ => 2);
            tryBottom.ToEither(Assert.IsType<BottomException>);
        }
```

**Comment by louthy** [comment]
Fixed in `v3.3.5`
